must-call-checker/README.md

```diff
@@ -0,0 +1,88 @@
+### MustCall Checker
+
+The MustCall Checker tracks the obligations that an object might have to call particular methods
+before it is deallocated. For example, an object whose type is `java.net.Socket` might
+have an obligation to call the `close()` method, to release the socket. Or, a builder object
+might have an obligation to call the `build()` method. We say that an object `obj` has a
+*must-call obligation* to call some method `foo` if there exists an execution path on which
+it is an error for `obj.foo()` not to be invoked before `obj` is deallocated. As a shorthand,
+the documentation for this checker sometimes shortens this to "`obj` must call `foo`" or similar,
+to refer to the worst case in which `obj` is actually an object that has to call `foo` (even if
+`obj` sometimes dynamically does not actually need to call `foo`, such as when `obj` is `null`).
+
+The MustCall Checker tracks which methods might need to be called, but does
+**not** check that these methods have actually been called.  The Object
+Construction Checker's `-AcheckMustCall` command-line option checks the
+obligation expressed by an `@MustCall` annotation. This option uses a dataflow
+algorithm to compare `@MustCall` obligations to inferred `@CalledMethods` facts,
+and reports cases where required methods might not have been called.
```

Comment:
> According to the documentation, no matter what @MustCall annotation appears on the type, the referent doesn't necessarily have any obligation to call any method

> Maybe the OCC is assuming a different semantics than expressed here, and the OCC assumes that @MustCall really means must call, as opposed to "might need to call".

I'll clarify this in the documentation, but the OCC isn't assuming a different semantics, but it does provide a different guarantee: the `-AcheckMustCall` command-line flag to the OCC guarantees that some set of must-call obligations were actually fulfilled. To that end, it's being conservative wrt what the MustCall checker can actually prove. If the MustCall checker can't prove that something is `@MustCall({})`, then the OCC assumes that the named methods really did need to be called, and checks them accordingly. Said another way, the MustCall checker's job is to discharge as many of these obligations as it can (by reducing them to `@MustCall({})` if possible), and the OCC checks the rest.

Reply:
That's very helpful, thanks.